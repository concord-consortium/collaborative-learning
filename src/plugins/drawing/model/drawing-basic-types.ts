import LineToolIcon from "../assets/line-icon.svg";
import SingleArrowIcon from "../assets/line-single-arrow-icon.svg";
import DoubleArrowIcon from "../assets/line-double-arrow-icon.svg";
import { FunctionComponent, SVGProps } from "react";

export interface Point { x: number; y: number; }

export interface BoundingBox {
  nw: Point;
  se: Point;
  start?: Point;
  end?: Point;
}

export interface BoundingBoxDelta {
  top: number, 
  right: number, 
  bottom: number, 
  left: number
}

export enum VectorType {
  line = "line",
  singleArrow = "arrow",
  doubleArrow = "doublearrow"
}

const vectorTypeIcons: Map<VectorType, FunctionComponent<SVGProps<SVGSVGElement>>> = new Map();
vectorTypeIcons.set(VectorType.line, LineToolIcon);
vectorTypeIcons.set(VectorType.singleArrow, SingleArrowIcon);
vectorTypeIcons.set(VectorType.doubleArrow, DoubleArrowIcon);
export function getVectorTypeIcon(vectorType?: VectorType) {
  return (vectorType && vectorTypeIcons.get(vectorType)) || LineToolIcon;
}

const vectorTypeTooltips: Map<VectorType, string> = new Map();
vectorTypeTooltips.set(VectorType.line, "Line");
vectorTypeTooltips.set(VectorType.singleArrow, "Arrow");
vectorTypeTooltips.set(VectorType.doubleArrow, "Double arrow");
export function getVectorTypeTooltip(vectorType?: VectorType) {
  return (vectorType && vectorTypeTooltips.get(vectorType)) || "Unknown";
}


// Possible decorations for the start and end of the vector.  Default is no decoration.
export enum VectorEndShape {
  triangle = "triangle"
}

// Return a two-element list of [head shape, tail shape] for the given VectorType constant.
export function endShapesForVectorType(vectorType?: VectorType) {
  if (!vectorType) {
    return [undefined, undefined];
  }
  switch (vectorType) {
    case VectorType.line:
      return [undefined, undefined];
    case VectorType.singleArrow:
      return [VectorEndShape.triangle, undefined];
    case VectorType.doubleArrow:
      return [VectorEndShape.triangle, VectorEndShape.triangle];
  }
}


export interface ToolbarSettings {
  stroke: string;
  fill: string;
  strokeDashArray: string;
  strokeWidth: number;
  vectorType?: VectorType;
}

export const DefaultToolbarSettings: ToolbarSettings = {
  stroke: "#000000",
  fill: "none",
  strokeDashArray: "",
  strokeWidth: 2,
  vectorType: VectorType.line
};
