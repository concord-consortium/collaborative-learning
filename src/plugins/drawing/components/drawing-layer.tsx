import React from "react";
import { reaction, IReactionDisposer } from "mobx";
import { clone, isAlive, getSnapshot } from "mobx-state-tree";
import { observer } from "mobx-react";
import { extractDragTileType, kDragTileContent } from "../../../components/tiles/tile-component";
import { DrawingContentModelType, DrawingObjectMove } from "../model/drawing-content";
import { ToolTileModelType } from "../../../models/tiles/tile-model";
import { safeJsonParse } from "../../../utilities/js-utils";
import { ImageContentSnapshotOutType } from "../../../models/tiles/image/image-content";
import { gImageMap } from "../../../models/image-map";
import { SelectionBox } from "./selection-box";
import { DrawingObjectSnapshotForAdd, DrawingObjectType, DrawingTool,
  HandleObjectHover,
  IDrawingLayer} from "../objects/drawing-object";
import { Point, ToolbarSettings } from "../model/drawing-basic-types";
import { getDrawingToolInfos, renderDrawingObject } from "./drawing-object-manager";
import { ImageObject } from "../objects/image";

const SELECTION_COLOR = "#777";
const HOVER_COLOR = "#bbdd00";
const SELECTION_BOX_PADDING = 10;

/**  ======= Drawing Layer ======= */
interface DrawingToolMap {
  [key: string]: DrawingTool | undefined;
}

interface DrawingLayerViewProps {
  model: ToolTileModelType;
  readOnly?: boolean;
  scale?: number;
  onSetCanAcceptDrop: (tileId?: string) => void;
  imageUrlToAdd?: string;
  setImageUrlToAdd?: (url: string) => void;
}

interface DrawingLayerViewState {
  toolbarSettings?: ToolbarSettings;
  currentDrawingObject: DrawingObjectType|null;
  selectedObjects: DrawingObjectType[];
  selectionBox: SelectionBox|null;
  hoverObject: DrawingObjectType|null;
  objectsBeingDragged: DrawingObjectType[];
}

@observer
export class DrawingLayerView extends React.Component<DrawingLayerViewProps, DrawingLayerViewState>
    implements IDrawingLayer {
  public currentTool: DrawingTool|null;
  public tools: DrawingToolMap;
  private svgRef: React.RefObject<any>|null;
  private setSvgRef: (element: any) => void;
  private _isMounted: boolean;
  private disposers: IReactionDisposer[];

  constructor(props: DrawingLayerViewProps) {
    super(props);

    this.state = {
      currentDrawingObject: null,
      selectionBox: null,
      selectedObjects: [],
      hoverObject: null,
      objectsBeingDragged: [],
    };

    this.tools = {};
    const drawingToolInfos = getDrawingToolInfos();
    drawingToolInfos.forEach(info => {
      if (info.toolClass) {
        this.tools[info.name] = new info.toolClass(this);
      }
    });

    this.currentTool = this.tools.select!;

    this.svgRef = null;
    this.setSvgRef = (element) => {
      this.svgRef = element;
    };
  }

  public componentDidMount() {
    this._isMounted = true;
    this.disposers = [];

    this.disposers.push(reaction(
      () => this.getContent().selectedButton,
      selectedButton => this.syncCurrentTool(selectedButton)
    ));

    // TODO: the list of selected objects is operated on directly by the model for example:
    // deleteSelectedObjects. So it is redundant to have the selection stored here in the state
    // too. There are still a few places working with this list of selected objects from the
    // state instead of the model.
    this.disposers.push(reaction(
      () => this.getContent().selectedIds,
      selectedIds => {
        const selectedObjects = selectedIds.map(
          id => this.getContent().objectMap[id]).filter(obj => !!obj) as DrawingObjectType[];
        this.setState({ selectedObjects });
      }
    ));

    this.disposers.push(reaction(
      () => this.getContent().toolbarSettings,
      settings => this.setCurrentToolSettings(settings)
    ));
  }

  public componentDidUpdate(prevProps: DrawingLayerViewProps, prevState: DrawingLayerViewState) {
    if (this.props.imageUrlToAdd) {
      this.addImage(this.props.imageUrlToAdd);
      this.props.setImageUrlToAdd?.("");
    }
  }

  public componentWillUnmount() {
    this.disposers.forEach(disposer => disposer());

    this._isMounted = false;
  }

  public syncCurrentTool(selectedButton: string) {
    const settings = this.getContent().toolbarSettings;
    const tool = this.tools[selectedButton];

    if (!tool) {
      console.warn("Unknown tool selected", selectedButton);
      return;
    }

    tool.setSettings(settings);
    this.setCurrentTool(tool);
  }

  public addNewDrawingObject(drawingObject: DrawingObjectSnapshotForAdd) {
    this.getContent().addObject(drawingObject);
  }

  public setSelectedObjects(selectedObjects: DrawingObjectType[]) {
    this.setState({selectionBox: null, selectedObjects});

    const drawingContent = this.props.model.content as DrawingContentModelType;
    const selectedObjectIds = selectedObjects.map(object => object.id || "");
    drawingContent.setSelection(selectedObjectIds);
  }

  public getSelectedObjects(): DrawingObjectType [] {
    return this.state.selectedObjects;
  }

  public setCurrentTool(tool: DrawingTool|null) {
    this.currentTool = tool;
  }

  public setCurrentToolSettings(settings: ToolbarSettings) {
    this.currentTool?.setSettings(settings);
  }

  public getCurrentStamp() {
    const drawingContent = this.props.model.content as DrawingContentModelType;
    return drawingContent.currentStamp;
  }

  public handleMouseDown = (e: React.MouseEvent<HTMLDivElement>) => {
    if (!this.props.readOnly && this.currentTool) {
      this.currentTool.handleMouseDown(e);
    }
  };

  public handleObjectClick = (e: MouseEvent|React.MouseEvent<any>, obj: DrawingObjectType) => {
    if (!this.props.readOnly && this.currentTool) {
      this.currentTool.handleObjectClick(e, obj);
    }
  };

  public handleObjectHover: HandleObjectHover = (e, obj, hovering) => {
    if (!this.props.readOnly && this.currentTool === this.tools.select) {
      this.setState({hoverObject: hovering ? obj : null});
    }
  };

  // handles dragging of selected/hovered objects
  // TODO: it seems this could be cleaned up. Keeping the state variable objectsBeingDragged
  // locally in this function and modifying the local variable is not safe.
  public handleSelectedObjectMouseDown = (e: React.MouseEvent<any>, obj: DrawingObjectType) => {
    if (this.props.readOnly || (this.currentTool !== this.tools.select)) return;
    let moved = false;
    const {selectedObjects, hoverObject } = this.state;
    let { objectsBeingDragged } = this.state;
    let objectsToInteract: DrawingObjectType[];
    let needToAddHoverToSelection = false;

    //If the object you are dragging is selected then the selection should not be cleared
    //and all objects should be moved.
    //If the object you are dragging was not selected then only then all of the other objects
    //should be deselected and just the object you are dragging should be selected.
    if (hoverObject && !selectedObjects.some(object => object.id === hoverObject.id)) {
      needToAddHoverToSelection = true;
      if (e.shiftKey || e.metaKey){
        objectsToInteract = [hoverObject, ...selectedObjects];
      }
      else {
        objectsToInteract = [hoverObject];
      }
    } else {
      objectsToInteract = selectedObjects;
    }

    const starting = this.getWorkspacePoint(e);
    if (!starting) return;

    e.stopPropagation();

    const handleMouseMove = (e2: MouseEvent) => {
      e2.preventDefault();
      e2.stopPropagation();

      const current = this.getWorkspacePoint(e2);
      if (!current) return;
      const dx = current.x - starting.x;
      const dy = current.y - starting.y;
      moved = moved || ((dx !== 0) || (dy !== 0));

      if (objectsBeingDragged.length === 0) {
        objectsBeingDragged = objectsToInteract.map(object => clone(object));
        this.setState( {objectsBeingDragged});
      }

      objectsBeingDragged.forEach((object, index) => {
        object.setPosition(objectsToInteract[index].x + dx, objectsToInteract[index].y + dy);
      });

      if (needToAddHoverToSelection) {
        // we delay until we confirm that the user is dragging the objects before adding the hover object
        // to the selection, to avoid messing with the click to select/deselect logic
        this.setSelectedObjects(objectsToInteract);
        // Note: the hoverObject could be kind of in a weird state here. It might
        // be both selected and hovered at the same time. However it is more
        // simple to keep the hoverObject independent of the selection. It just
        // represents the current object the mouse is over regardless of whether
        // it is selected or not.
        needToAddHoverToSelection = false;
      }
    };
    const handleMouseUp = (e2: MouseEvent) => {
      e2.preventDefault();
      e2.stopPropagation();
      window.removeEventListener("mousemove", handleMouseMove);
      window.removeEventListener("mouseup", handleMouseUp);
      if (moved) {
        const moves: DrawingObjectMove[] = objectsToInteract.map((object, index) => {
          const draggedObject = objectsBeingDragged[index];
          return {
            id: object.id,
            destination: {x: draggedObject.x, y: draggedObject.y}
          };
        });
        this.getContent().moveObjects(moves);
      }
      else {
        this.handleObjectClick(e2, obj);
      }
      this.setState({ objectsBeingDragged: []});
    };

    window.addEventListener("mousemove", handleMouseMove);
    window.addEventListener("mouseup", handleMouseUp);
  };

  public startSelectionBox(p: Point) {
    this.setState({selectionBox: new SelectionBox(p)});
  }

  public updateSelectionBox(p: Point) {
    const {selectionBox} = this.state;
    if (selectionBox) {
      selectionBox.update(p);
      this.setState({selectionBox});
    }
  }

  public endSelectionBox(addToSelectedObjects: boolean) {
    const {selectionBox} = this.state;
    if (selectionBox) {
      selectionBox.close();
      const selectedObjects: DrawingObjectType[] = addToSelectedObjects ? this.state.selectedObjects : [];
      this.forEachObject((object) => {
        if (object.inSelection(selectionBox)) {
          if (selectedObjects.indexOf(object) === -1) {
            selectedObjects.push(object);
          }
        }
      });
      this.setSelectedObjects(selectedObjects);
    }
  }

  // when we add text, this filter can be used with this.renderObjects((object) => object.type !== "text")
  public renderObjects(_filter: (object: DrawingObjectType) => boolean) {
    return this.getContent().objects.map((object) => {
      if (!object || !_filter(object)) {
        return null;
      }
      return renderDrawingObject(object, this.handleObjectHover, this.handleSelectedObjectMouseDown);
    });
  }

  public renderSelectedObjects(selectedObjects: DrawingObjectType[], color: string) {
    return selectedObjects.map((object, index) => {
      let {nw: {x: nwX, y: nwY}, se: {x: seX, y: seY}} = object.boundingBox;
      nwX -= SELECTION_BOX_PADDING;
      nwY -= SELECTION_BOX_PADDING;
      seX += SELECTION_BOX_PADDING;
      seY += SELECTION_BOX_PADDING;
      return <rect
                key={index}
                data-testid="selection-box"
                x={nwX}
                y={nwY}
                width={seX - nwX}
                height={seY - nwY}
                fill={color}
                fillOpacity="0"
                stroke={color}
                strokeWidth="1.5"
                strokeDasharray="10 5"
                pointerEvents={"none"}
               />;
    });
  }

  //we want to populate our objectsBeingDragged state array

  public setCurrentDrawingObject(object: DrawingObjectType | null) {
    this.setState({currentDrawingObject: object});
  }

  public getWorkspacePoint = (e: MouseEvent|React.MouseEvent<any>): Point|null => {
    if (this.svgRef) {
      const scale = this.props.scale || 1;
      const rect = ((this.svgRef as unknown) as Element).getBoundingClientRect();
      return {
        x: (e.clientX - rect.left) / scale,
        y: (e.clientY - rect.top) / scale
      };
    }
    return null;
  };

  public render() {
    const hoveringOverAlreadySelectedObject =
      this.state.hoverObject
        ? this.state.selectedObjects.indexOf(this.state.hoverObject) !== -1
        : false;

    const idsBeingDragged = this.state.objectsBeingDragged.map(object => object.id);
    const objectsToRenderSelected = this.state.objectsBeingDragged.length > 0 ?
      this.state.objectsBeingDragged : this.state.selectedObjects;

    return (
      <div className="drawing-layer"
          data-testid="drawing-layer"
          onMouseDown={this.handleMouseDown}
          onDragOver={this.handleDragOver}
          onDragLeave={this.handleDragLeave}
          onDrop={this.handleDrop} >

        <svg xmlnsXlink="http://www.w3.org/1999/xlink" width={1500} height={1500} ref={this.setSvgRef}>
          {this.renderObjects(object => object.type === "image" && !idsBeingDragged.includes(object.id))}
          {this.renderObjects(object => object.type !== "image" && !idsBeingDragged.includes(object.id))}
          {this.state.objectsBeingDragged.map((object) => renderDrawingObject(object))}
          {this.renderSelectedObjects(objectsToRenderSelected, SELECTION_COLOR)}
          {(this.state.hoverObject && !hoveringOverAlreadySelectedObject && isAlive(this.state.hoverObject))
            ? this.renderSelectedObjects([this.state.hoverObject], HOVER_COLOR)
            : null}
          {this.state.currentDrawingObject
            ? renderDrawingObject(this.state.currentDrawingObject)
            : null}
          {this.state.selectionBox ? this.state.selectionBox.render() : null}
        </svg>
      </div>
    );
  }

  private getContent() {
    return this.props.model.content as DrawingContentModelType;
  }

  private isAcceptableImageDrag = (e: React.DragEvent<HTMLDivElement>) => {
    const { readOnly } = this.props;
    const toolType = extractDragTileType(e.dataTransfer);
    // image drop area is central 80% in each dimension
    const kImgDropMarginPct = 0.1;
    if (!readOnly && (toolType === "image")) {
      const eltBounds = e.currentTarget.getBoundingClientRect();
      const kImgDropMarginX = eltBounds.width * kImgDropMarginPct;
      const kImgDropMarginY = eltBounds.height * kImgDropMarginPct;
      if ((e.clientX > eltBounds.left + kImgDropMarginX) &&
          (e.clientX < eltBounds.right - kImgDropMarginX) &&
          (e.clientY > eltBounds.top + kImgDropMarginY) &&
          (e.clientY < eltBounds.bottom - kImgDropMarginY)) {
        return true;
      }
    }
    return false;
  };

  private handleDragOver = (e: React.DragEvent<HTMLDivElement>) => {
    const isAcceptableDrag = this.isAcceptableImageDrag(e);
    // TODO: what is this method used for?
    this.props.onSetCanAcceptDrop(isAcceptableDrag ? this.props.model.id : undefined);
    if (isAcceptableDrag) {
      e.dataTransfer.dropEffect = "copy";
      e.preventDefault();
    }
  };

  private handleDragLeave = (e: React.DragEvent<HTMLDivElement>) => {
    this.props.onSetCanAcceptDrop();
  };

  private handleDrop = (e: React.DragEvent<HTMLDivElement>) => {
    if (this.isAcceptableImageDrag(e)) {
      const dragContent = e.dataTransfer.getData(kDragTileContent);
      const parsedContent = safeJsonParse(dragContent);
      if (parsedContent) {
        const droppedContent: ImageContentSnapshotOutType = parsedContent.content;
        const droppedUrl = droppedContent.url;
        if (droppedUrl) {
          this.handleImageDrop(droppedUrl);
        }
        e.preventDefault();
        e.stopPropagation();
      }
    }
  };

  private handleImageDrop(url: string) {
    this.props.setImageUrlToAdd?.(url);
  }

  private addImage(url: string) {
    gImageMap.getImage(url)
      .then(imageEntry => {
        if (!this._isMounted || !imageEntry.contentUrl) return;
        const image = ImageObject.create({
          // The contentUrl is used here because that is the normalized URL for this
          // image.
          url: imageEntry.contentUrl,
          // The filename is stored because the imageEntry is temporary and the filename
          // cannot be recreated just from the URL.  In this case, the imageEntry is being
          // used to transport this filename from the source of the drop.
          filename: imageEntry.filename,
          x: 0,
          y: 0,
          width: imageEntry.width!,
          height: imageEntry.height!
        });
        this.addNewDrawingObject(getSnapshot(image));
      });
  }

  private forEachObject(callback: (object: DrawingObjectType, key?: string) => void) {
    const {objects} = this.getContent();
    objects.forEach((object) => {
      if (object) {
        callback(object, object.id);
      }
    });
  }
}
