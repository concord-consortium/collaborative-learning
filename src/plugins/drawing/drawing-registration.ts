import { registerTileComponentInfo } from "../../models/tiles/tile-component-info";
import { registerTileContentInfo } from "../../models/tiles/tile-content-info";
import { DrawingContentModel, DrawingToolMetadataModel, defaultDrawingContent } from "./model/drawing-content";
import { kDrawingToolID, kDrawingDefaultHeight } from "./model/drawing-types";
import DrawingToolComponent from "./components/drawing-tile";
import DrawingToolIcon from "../../clue/assets/icons/draw-tool.svg";
import { DrawingMigrator } from "./model/drawing-migrator";

registerTileContentInfo({
  id: kDrawingToolID,
  titleBase: "Drawing",
  // TODO: maybe there is a better way to do this kind of casting?
  //   The issue is that modelClass prop has a type of `typeof TileContentModel`,
  //   That type is pretty restrictive and doesn't accommodate the return of
  //   types.snapshotProcessor. There might be a better way to get a
  //   typescript type for a MST "Class" which is less restrictive
  modelClass: DrawingMigrator as typeof DrawingContentModel,
  metadataClass: DrawingToolMetadataModel,
  defaultHeight: kDrawingDefaultHeight,
  exportNonDefaultHeight: true,
  defaultContent: defaultDrawingContent
});

registerTileComponentInfo({
  id: kDrawingToolID,
  Component: DrawingToolComponent,
  tileEltClass: "drawing-tool-tile",
  Icon: DrawingToolIcon
});
