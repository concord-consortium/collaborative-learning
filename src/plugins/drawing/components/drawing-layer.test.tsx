import { render, screen } from "@testing-library/react";
import React from "react";
import { TileModel } from "../../../models/tiles/tile-model";
import { createDrawingContent, DrawingContentModelType } from "../model/drawing-content";
import { DrawingLayerView } from "./drawing-layer";
import { LineObject, LineObjectType } from "../objects/line";
import { VectorObject, VectorObjectType } from "../objects/vector";
import { VectorEndShape, VectorType, endShapesForVectorType } from "../model/drawing-basic-types";
import { RectangleObject, RectangleObjectSnapshotForAdd, RectangleObjectType } from "../objects/rectangle";
import { EllipseObject, EllipseObjectType } from "../objects/ellipse";
import { ImageObject, ImageObjectType } from "../objects/image";

// The drawing tile needs to be registered so the TileModel.create
// knows it is a supported tile type
import "../drawing-registration";

let content: DrawingContentModelType, drawingLayerProps, drawingLayer;

const kLocalImageUrl = "assets/logo_tw.png";

const getDrawingObject = (objectContent: DrawingContentModelType) => {
  drawingLayerProps = {
    model: TileModel.create({content: objectContent}),
    onSetCanAcceptDrop: (tileId?: string) => {
      throw new Error("Function not implemented.");
    }
  };
  render(<DrawingLayerView {...drawingLayerProps} />);
  drawingLayer = screen.getByTestId("drawing-layer");
  return drawingLayer.firstChild;
};

describe("Drawing Layer Components", () => {
  describe("Freehand Line", () => {
    let line: LineObjectType;
    beforeEach(() => {
      line = LineObject.create({
        id: "123",
        x: 10, y: 10,
        deltaPoints: [{ dx: 1, dy: 1 }, { dx: 2, dy: 2 }],
        stroke: "#888888",
        strokeDashArray: "3,3",
        strokeWidth: 1
      });
      content = createDrawingContent({ objects: [line] });
    });

    it("adds a freehand line", () => {
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a freehand line", () => {
      line.setPosition(5, 5);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("deletes a freehand line", () => {
      content.deleteObjects([line.id]);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });

  describe("Vector line", () => {
    let vector: VectorObjectType;
    beforeEach(() => {
      vector = VectorObject.create({
        id: "234",
        x: 10, y: 10,
        dx: 10, dy: 10,
        stroke: "#888888",
        strokeDashArray: "3,3",
        strokeWidth: 1
      });
      content = createDrawingContent({ objects: [vector] });
    });
    it("adds a Vector line", () => {
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a vector line", () => {
      vector.setPosition(5,5);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("deletes a vector line", () => {
      content.deleteObjects([vector.id]);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });

  describe("Vector arrow", () => {
    let vector: VectorObjectType;
    beforeEach(() => {
      vector = VectorObject.create({
        id: "235",
        x: 20, y:20,
        dx:20, dy:20,
        stroke: "#777777",
        strokeDashArray: "1,1",
        strokeWidth: 2,
        headShape: VectorEndShape.triangle,
        tailShape: VectorEndShape.triangle
      });
      content = createDrawingContent({ objects: [vector] });
    });
    it("adds a Vector arrow", () => {
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a vector arrow", () => {
      vector.setPosition(5,5);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("changes vector to single arrow", () => {
      vector.setEndShapes(...endShapesForVectorType(VectorType.singleArrow));
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("changes vector to double arrow", () => {
      vector.setEndShapes(...endShapesForVectorType(VectorType.doubleArrow));
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("deletes a vector arrow", () => {
      content.deleteObjects([vector.id]);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });

  describe("Rectangle", () => {
    let rect: RectangleObjectType;
    beforeEach(() => {
      rect = RectangleObject.create({
        id: "345",
        x: 10, y: 10,
        width: 10, height: 10,
        fill: "#cccccc",
        stroke: "#888888",
        strokeDashArray: "3,3",
        strokeWidth: 1
      });
      content = createDrawingContent({ objects: [rect] });
    });
    it("adds a Rectangle", () => {
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a rectangle", () => {
      rect.setPosition(5,5);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("deletes a rectangle", () => {
      content.deleteObjects([rect.id]);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });

  describe("Ellipse", () => {
    let ellipse: EllipseObjectType;
    beforeEach(() => {
      ellipse = EllipseObject.create({
        id: "456",
        x: 10, y: 10,
        rx: 10, ry: 10,
        fill: "#cccccc",
        stroke: "#888888",
        strokeDashArray: "3,3",
        strokeWidth: 1
      });
      content = createDrawingContent({ objects: [ellipse] });
    });
    it("adds a ellipse", () => {
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a ellipse", () => {
      ellipse.setPosition(5,5);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("deletes a ellipse", () => {
      content.deleteObjects([ellipse.id]);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });

  // TODO: should test TextObject, but it fails
  // because the virual DOM doesn't implement getComputedTextLength

  describe("Group", () => {
    let rectangle: RectangleObjectType;
    let ellipse: EllipseObjectType;
    beforeEach(() => {
      rectangle = RectangleObject.create({
        id: "r",
        x: 10, y: 10,
        width: 10, height: 10,
        fill: "#cccccc",
        stroke: "#888888",
        strokeDashArray: "3,3",
        strokeWidth: 1
      });
      ellipse = EllipseObject.create({
        id: "e",
        x: 10, y: 10,
        rx: 10, ry: 10,
        fill: "#cccccc",
        stroke: "#888888",
        strokeDashArray: "3,3",
        strokeWidth: 1
      });
      content = createDrawingContent({ objects: [rectangle, ellipse] });
    });
    it("creates a group", () => {
      content.createGroup(["r", "e"]);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a group", () => {
      content.createGroup(["r", "e"]);
      expect(content.objects[0].type).toBe("group");
      content.objects[0].setDragPosition(5, 5);
      content.objects[0].repositionObject();
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("ungroups a group", () => {
      content.createGroup(["r", "e"]);
      const r2: RectangleObjectSnapshotForAdd = {
        type: "rectangle",
        id: "not-in-group",
        x: 10, y: 10,
        width: 10, height: 10,
        fill: "#cccccc",
        stroke: "#888888",
        strokeDashArray: "3,3",
        strokeWidth: 1
      };
      content.addObject(r2);
      content.ungroupGroups([content.objects[0].id, "not-in-group"]);
      expect(content.objects.length).toBe(3);
      expect(content.objects.map((obj) => obj.id)).toStrictEqual(["not-in-group", "r", "e"]);
    });
    it("deletes a group", () => {
      content.createGroup(["r", "e"]);
      content.deleteObjects([content.objects[0].id]);
      expect(content.objects.length).toBe(0);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });

  describe("Image", () => {
    let image: ImageObjectType;
    beforeEach(() => {
      image = ImageObject.create({
        id: "567",
        url: kLocalImageUrl,
        x: 10, y: 10,
        width: 10, height: 10,
      });
      content = createDrawingContent({ objects: [image] });
    });
    it("adds an image", () => {
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a image", () => {
      image.setPosition(5,5);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("deletes a image", () => {
      content.deleteObjects([image.id]);
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });
});

describe("Drawing Layer Zoom", () => {
  let rect: RectangleObjectType;
  beforeEach(() => {
    rect = RectangleObject.create({
      id: "345",
      x: 10, y: 10,
      width: 10, height: 10,
      fill: "#cccccc",
      stroke: "#888888",
      strokeDashArray: "3,3",
      strokeWidth: 1
    });
    content = createDrawingContent({ objects: [rect] });
  });

  it("zooms in", () => {
    content.setZoom(2);
    expect(getDrawingObject(content)).toMatchSnapshot();
  });
  it("zooms out", () => {
    content.setZoom(0.5);
    expect(getDrawingObject(content)).toMatchSnapshot();
  });
});
