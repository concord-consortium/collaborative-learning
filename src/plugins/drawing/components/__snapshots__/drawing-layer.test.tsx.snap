// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawing Layer Components Ellipse adds a ellipse 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <ellipse
    cx="10"
    cy="10"
    fill="#cccccc"
    pointer-events="visible"
    rx="10"
    ry="10"
    stroke="#888888"
    stroke-dasharray=""
    stroke-width="1"
  />
</svg>
`;

exports[`Drawing Layer Components Ellipse deletes a ellipse 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
/>
`;

exports[`Drawing Layer Components Ellipse moves a ellipse 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <ellipse
    cx="5"
    cy="5"
    fill="#cccccc"
    pointer-events="visible"
    rx="10"
    ry="10"
    stroke="#888888"
    stroke-dasharray=""
    stroke-width="1"
  />
</svg>
`;

exports[`Drawing Layer Components Freehand Line adds a freehand line 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M 10 10 l 1 1 l 2 2"
    fill="none"
    pointer-events="visible"
    stroke="#888888"
    stroke-dasharray=""
    stroke-width="1"
  />
</svg>
`;

exports[`Drawing Layer Components Freehand Line deletes a freehand line 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
/>
`;

exports[`Drawing Layer Components Freehand Line moves a freehand line 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M 5 5 l 1 1 l 2 2"
    fill="none"
    pointer-events="visible"
    stroke="#888888"
    stroke-dasharray=""
    stroke-width="1"
  />
</svg>
`;

exports[`Drawing Layer Components Image adds an image 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <image
    height="10"
    href="test-file-stub"
    preserveAspectRatio="none"
    width="10"
    x="10"
    y="10"
  />
</svg>
`;

exports[`Drawing Layer Components Image deletes a image 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
/>
`;

exports[`Drawing Layer Components Image moves a image 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <image
    height="10"
    href="assets/logo_tw.png"
    preserveAspectRatio="none"
    width="10"
    x="5"
    y="5"
  />
</svg>
`;

exports[`Drawing Layer Components Rectangle adds a Rectangle 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <rect
    fill="#cccccc"
    height="10"
    pointer-events="visible"
    stroke="#888888"
    stroke-dasharray=""
    stroke-width="1"
    width="10"
    x="10"
    y="10"
  />
</svg>
`;

exports[`Drawing Layer Components Rectangle deletes a rectangle 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
/>
`;

exports[`Drawing Layer Components Rectangle moves a rectangle 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <rect
    fill="#cccccc"
    height="10"
    pointer-events="visible"
    stroke="#888888"
    stroke-dasharray=""
    stroke-width="1"
    width="10"
    x="5"
    y="5"
  />
</svg>
`;

exports[`Drawing Layer Components Vector line adds a Vector line 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <line
    stroke="#888888"
    stroke-dasharray=""
    stroke-width="1"
    x1="10"
    x2="20"
    y1="10"
    y2="20"
  />
</svg>
`;

exports[`Drawing Layer Components Vector line deletes a vector line 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
/>
`;

exports[`Drawing Layer Components Vector line moves a vector line 1`] = `
<svg
  height="1500"
  width="1500"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <line
    stroke="#888888"
    stroke-dasharray=""
    stroke-width="1"
    x1="5"
    x2="15"
    y1="5"
    y2="15"
  />
</svg>
`;
